web/src/lib/components/asset-viewer/editor/crop-canvas.svelte

```diff
@@ -0,0 +1,740 @@
+<script lang="ts">
+  import {
+    cropAspectRatio,
+    cropImageScale,
+    cropImageSize,
+    cropSettings,
+    type CropSettings,
+    type CropAspectRatio,
+    cropSettingsChanged,
+  } from '$lib/stores/asset-editor.store';
+  import { onMount, afterUpdate } from 'svelte';
+  import { get } from 'svelte/store';
+
+  export let crop = get(cropSettings);
+  export let asset;
+  export let aspectRatio = get(cropAspectRatio);
+  cropSettings.subscribe((value) => {
+    crop = value;
+  });
+
+  let canvas: HTMLCanvasElement;
+  let ctx: CanvasRenderingContext2D;
+  let img: HTMLImageElement;
+
+  let darkenLevel = 0.65; // Initial darkening level
+  let animationFrame: ReturnType<typeof requestAnimationFrame>;
+  let canvasCursor: string;
+  const getAssetUrl = (id: string, checksum: string) => {
+    return `http://localhost:2283/api/assets/${id}/original?c=${checksum}`;
+  };
+
+  const draw = () => {
+    if (!ctx || !canvas) {
+      return;
+    }
+    cropSettings.set(crop);
+    ctx.clearRect(0, 0, canvas.width, canvas.height);
+    ctx.drawImage(img, 0, 0, canvas.width, canvas.height);
+    drawOverlay();
+    drawCropRect();
+  };
+
+  cropAspectRatio.subscribe((value) => {
+    aspectRatio = value;
+    const newCrop = recalculateCrop(true); // Recalculate the crop when the aspect ratio changes and return the new value
+    if (newCrop) {
+      animateCropChange(newCrop);
+    }
+  });
+
+  const drawCropRect = () => {
+    ctx.globalCompositeOperation = 'exclusion';
+    ctx.strokeStyle = 'white';
+    ctx.lineWidth = 2;
+    ctx.strokeRect(crop.x, crop.y, crop.width, crop.height);
+
+    // Set blend mode
+    ctx.strokeStyle = 'white';
+    ctx.lineWidth = 1;
+
+    // Vertical lines
+    const thirdWidth = crop.width / 3;
```

Comment:
Not a huge fan of the grid lines personally, maybe only show them while changing the crop area?

Reply:
I made the lines visible only when the cropping rectangle is changed. But then, when the cropping rectangle is the same size as the image, it is poorly visible
![image](https://github.com/user-attachments/assets/21e75fbd-9bc4-411a-9e86-8cb2e5fc3dae)
I can't imagine how to solve this in a simple way. I added small circles on the corners for more visibility. 
![image](https://github.com/user-attachments/assets/d8bfbe95-912b-440c-9df6-85a4957bdcf2)
But maybe as long as the rectangle is the same size as the image, draw these lines?
Although if this is just a problem for me, then everything is ok.

Reply:
Yeah I think this is better. The dots only seem to get cut off at the edges, maybe there's a way to extend them? I also noticed that cropping immediately stops when the cursor leaves the canvas. Can we allow cropping to continue when that happens?

Reply:
It will require some reworking, but I will try

Reply:
It wasn't easy, but it's done. [video](https://um.ichuk.ru/s/miEkaAbmzmTJ2Fs)

Reply:
Another thing that might be worth looking into down the road is using sharp to do the actual cropping on the original asset. So always show the crop tool on the jpeg/webp thumbnail and use it to get the cropping coordinates, but when send those to the server which can actually perform the operation. I believe using sharp it would be possible to keep the metadata, etc.